captum/attr/_core/layer/layer_lrp.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+import torch
+
+from ..._core.lrp import LRP
+from ..._utils.attribution import LayerAttribution
+from ..._utils.common import _format_attributions, _format_input
+from ..._utils.gradient import (
+    _forward_layer_eval,
+    apply_gradient_requirements,
+    compute_gradients,
+    undo_gradient_requirements,
+)
+
+
+class LayerLRP(LRP, LayerAttribution):
+    r"""
+    Layer-wise relevance propagation is based on a backward propagation
+    mechanism applied sequentially to all layers of the model. Here, the
+    model output score represents the initial relevance which is decomposed
+    into values for each neuron of the underlying layers. The decomposition
+    is defined by rules that are chosen for each layer, involving its weights
+    and activations. Details on the model can be found in the original paper
+    [https://doi.org/10.1371/journal.pone.0130140]. The implementation is
+    inspired by the tutorial of the same group
+    [https://doi.org/10.1016/j.dsp.2017.10.011] and the publication by
+    Ancona et al. [https://openreview.net/forum?id=Sy21R9JAW].
+
+    Warning: In-place relus lead to unexptected failures in layer LRP.
```

Comment:
In layerLRP were you setting any hooks on the ReLUs or accessing the input of the Relu after running forward pass on it ?

nit: `unexpected`

Reply:
Running the commented test case `test_layer_lrp.test_lrp_simple_inplace` and printing the input relevances and the output and input from the ReLUs gives strange outputs:

```
relevance: tensor([0.3333, 0.3333, 0.3333])
module:ReLU() output:(tensor([0.3333, 0.3333, 0.3333]),) input: (tensor([0.3333, 0.3333, 0.3333]),)
relevance: tensor([0.1667, 0.3333, 0.5000])
module:ReLU(inplace=True) output:(tensor([0.0278, 0.0278, 0.0278]),) input: (tensor([0.0278, 0.0278, 0.0278]),)
relevance: tensor([0.0278, 0.0278, 0.0278])
relevance: tensor([0.1667, 0.3333, 0.5000]
```

It appears as if the order in which the hooks are called is different if inplace ReLUs are used. I cannot understand where this is coming from.

Reply:
I'll try to debug this as well. In-place Relus are pretty common in pre-trained models. e.g. in inception model.

Reply:
@nanohanno, to solve the problem with inplace relus we need to return  `return outputs.clone()` from `forward_hook` in lrp_rules.py.
The problem is that with inplace relus our inputs are being overwritten with the outputs and that's why the ordering also looks weird.